google-cloud-dialogflow/MIGRATING.md

```diff
@@ -0,0 +1,274 @@
+    request object instead of separate arguments. See
+    [Passing Arguments](#passing-arguments) for more info.
+ *  Previously, some client classes included class methods for constructing
+    resource paths. These paths are now instance methods, and are also
+    available in a separate paths module. See
+    [Resource Path Helpers](#resource-path-helpers) for more info.
+ *  Some classes have moved into different namespaces. See
+    [Class Namespaces](#class-namespaces) for more info.
+
+### Library Structure
+
+Older 0.x releases of the `google-cloud-dialogflow` gem were all-in-one gems that
+included potentially multiple clients for multiple versions of the Dialogflow
+service. Factory methods such as `Google::Cloud::Dialogflow::Agents.new()` would
+return you client instances such as `Google::Cloud::Dialogflow::V2::AgentsClient`.
+These classes were all defined in the same gem.
+
+With the 1.0 release, the `google-cloud-dialogflow` gem still provides factory
+methods for obtaining clients. (The method signatures will have changed. See
+[Creating Clients](#creating-clients) for details.) However, the actual client
+classes have been moved into separate gems, on per service version. Currently,
+Dialogflow has one version, V2. The
+`Google::Cloud::Dialogflow::V2::Agents::Client` class, along with its
+helpers and data types, are now part of the `google-cloud-dialogflow-v2` gem.
+
+For normal usage, you can continue to install the `google-cloud-dialogflow` gem
+and continue to use factory methods to create clients. This will remain the
+easiest way to use the Dialogflow client. However, you may alternatively
+choose to install only one of the versioned gems. For example, if you know you
+will only use `V2` of the service, you can install `google-cloud-dialogflow-v2`
+by itself, and construct instances of the
+`Google::Cloud::Dialogflow::V2::Agents::Client` client class directly.
+
+### Client Configuration
+
+In older releases, if you wanted to customize performance parameters or
+low-level behavior of the client (such as credentials, timeouts, or
+instrumentation), you would pass a variety of keyword arguments to the client
+constructor. It was also extremely difficult to customize the default settings.
+
+With the 1.0 release, a configuration interface provides access to these
+parameters, both global defaults and per-client settings. For example, to set
+global credentials and default timeout for all Dialogflow V2 sessions clients:
+
+```
+Google::Cloud::Dialogflow::V2::Sessions::Client.configure do |config|
+  config.credentials = "/path/to/credentials.json"
+  config.timeout = 10_000
+end
+```
+
+Individual RPCs can also be configured independently. For example, to set the
+timeout for the `detect_intent` call:
+
+```
+Google::Cloud::Dialogflow::V2::Sessions::Client.configure do |config|
+  config.rpcs.detect_intent.timeout = 20_000
+end
+```
+
+You can also set certain configuration defaults for all Dialogflow versions and
```

Comment:
Why not lead with the broadest scope first? Then the progression would be consistently general -> specific.

Reply:
The ordering is intentional. It leads with the most straightforward and common cases, and moves toward less common cases.